File: trunk/ShadowFrameworkLite20/src/shadow/renderer/data/SFProgramAssetData.java
package shadow.renderer.data;

import shadow.renderer.SFProgramModuleStructures;
import shadow.system.data.SFNamedParametersObject;
import shadow.system.data.objects.SFString;

public class SFProgramAssetData extends SFDataAsset<SFProgramModuleStructures>{

	protected SFString program = new  SFString("");
	protected SFDataObjectsList<SFStructureReferenceDataObject> structures =new SFDataObjectsList<SFStructureReferenceDataObject>(new SFStructureReferenceDataObject());
	private SFDataObjectsList<SFTextureData> usedTexture = new SFDataObjectsList<SFTextureData>(new SFTextureData());

	public SFProgramAssetData(){
		SFNamedParametersObject parameters=new SFNamedParametersObject();
		parameters.addObject("program", program);
		parameters.addObject("structures", structures);
		parameters.addObject("textures", usedTexture);
		setData(parameters);
	}
	
	public void setProgram(String programName){
		program.setString(programName);
	}

	public void addStructure(String tableName,int index) {
		SFStructureReferenceDataObject object=new SFStructureReferenceDataObject();
		object.setRefIndex(index);
		object.setTableName(tableName);
		this.structures.add(object);
	}
	

	public void addTexture(String textureName) {
		SFTextureData object=new SFTextureData();
		object.getReference().setReference(textureName);
		this.usedTexture.add(object);
	}
	
	public void addTexture(int index, String textureSet) {
		SFTextureData texture = new SFTextureData();
		texture.setTextureIndex(index);
		texture.getReference().setReference(textureSet);
		usedTexture.add(texture);
	}
	
	@Override
	protected SFProgramModuleStructures buildResource() {
		SFProgramModuleStructures element=new SFProgramModuleStructures();
		
		element.setProgram(program.getString());
		
		for (SFStructureReferenceDataObject structure:structures) {
			element.getData().add(structure.buildReference());
		}
		
		//texturesCount=0;
		for (int i = 0; i < usedTexture.size(); i++) {
			element.getTextures().add(usedTexture.get(i).buildTextureReference());
		}
		
		return element;
	}
}

File: ShadowFrameworkJS/src/shadow/renderer/data/SFProgramAssetData.js


function SFProgramAssetData(){
	this.initialize();
}

inherit(SFProgramAssetData,SFDataAsset);

SFProgramAssetData.prototype["initialize"]=function(){
			this.program= new  SFString("");
			this.structures= new  SFDataObjectsList(new SFStructureReferenceDataObject());
			this.usedTexture= new  SFDataObjectsList(new SFTextureData());
			var parameters=new SFNamedParametersObject();
			parameters.addObject( this.program);
			parameters.addObject( this.structures);
			parameters.addObject( this.usedTexture);
			this.setData(parameters);
		};

SFProgramAssetData.prototype["buildResource"]=function(){
			var element=new SFProgramModuleStructures();
			
			element.setProgram(this.program.getString());
			
			for(var i=0;i<this.structures.size();i++){
				element.getData().push(this.structures.get(i).buildReference());
			}
			
			//texturesCount=0;
			for (var i = 0; i < this.usedTexture.size(); i++) {
				element.getTextures().push(this.usedTexture.get(i).buildTextureReference());
			}
			
			return element;
		};
		
SFProgramAssetData.prototype["generateNewDatasetInstance"]=function(){
	return new SFProgramAssetData();
};